stats/system_tag.go

```diff
@@ -0,0 +1,132 @@
+package stats
+
+import (
+	"bytes"
+	"encoding/json"
+	"strings"
+)
+
+// SystemTagSet is a bitmask that is used to keep track
+// which system tags should be included with which metrics.
+//go:generate enumer -type=SystemTagSet -transform=snake -trimprefix=Tag -output system_tag_set_gen.go
+type SystemTagSet uint32
+
+// TagSet is a string to bool map (for lookup efficiency) that is used to keep track
+// which system tags and non-system tags should be included with with metrics.
+type TagSet map[string]bool
```

Comment:
This can be a `map[string]struct{}` and it will save some memory .... probably not a lot but still

Reply:
Then how can you blacklist one?

Reply:
`_, ok := TagSet["tag1"]` 
is the same as 
`ok := TagSet["tag1"]`
 if all you set to is `true` and if you are using the `TagSet` to set `false` you can just ... not set it at all or delete it from the map.

Reply:
Ah, so we currently blacklist by not set it, instead of .setting it to `false`?

Reply:
What @MStoykov is referring to is that the `TagSet` is just meant to represent set (in the sense of a mathematical set) of unique strings (the tag names), so the important part of this is `map[string]`, not what comes after. And `map[string]bool` takes a few more bytes than `map[string]struct{}`...

That said, I disagree with @MStoykov - we only have a few `TagSet` instances per test run, so a few extra bytes don't make a difference. On the other hand, with the current type, the syntax of `if someTagSet[tagName] { /* do something */ }` is much more readable than `if _, ok := someTagSet[tagName]; ok { /* do something */ }`...

Reply:
The current condition:

```
if value != "" && !t[key] {
```

So the key is matter, no?

Reply:
@cuonglm, sorry, I don't understand what you mean. Can you elaborate?

Reply:
I would argue that TagSet could get a Has method like: 
```golang
// Has checks a tag included in tag set.
func (i TagSet) Has(tag string) bool {
	if i == nil {
		return false
	}
        _, ok := i[tag] 
        return ok
}
```

Reply:
@na-- The current usage of `lib.TagSet` https://github.com/loadimpact/k6/blob/4f04c7fa523ca46c9d48b3407d82d3edf33815cb/stats/datadog/collector.go#L34

As I understand, we currently use `map[string]bool` because we can blacklist a tag by set it to `false`.

Reply:
@na-- @MStoykov Oh, so if tag exists in map, then it will be blacklist right? If yes, then `map[string]struct{}` is the way to go.

Reply:
That is true but ... I found an [old benchmark](https://gist.github.com/davecheney/3be245c92b61e5045f75) and update it to allocate the maps on each iteration, the results are not flattering for: 
```
BenchmarkMapBool-4       1000000              1284 ns/op             717 B/op          1 allocs/op
BenchmarkMapBool-4       1000000              1289 ns/op             717 B/op          1 allocs/op
BenchmarkMapBool-4       1000000              1297 ns/op             717 B/op          1 allocs/op
BenchmarkMapBool-4       1000000              1334 ns/op             717 B/op          1 allocs/op
BenchmarkMapBool-4       1000000              1272 ns/op             717 B/op          1 allocs/op
BenchmarkMapStruct-4     1000000              1257 ns/op             646 B/op          1 allocs/op
BenchmarkMapStruct-4     1000000              1244 ns/op             646 B/op          1 allocs/op
BenchmarkMapStruct-4     1000000              1316 ns/op             645 B/op          1 allocs/op
BenchmarkMapStruct-4     1000000              1254 ns/op             645 B/op          1 allocs/op
BenchmarkMapStruct-4     1000000              1250 ns/op             645 B/op          1 allocs/op
```
the map[string]struct{} is consistently taking more CPU time and the amount of Bytes saved per map are so small that is not worth it ... once I've benchmarked it. I was expecting also some ... CPU boost but it's totally the opposite:
```
BenchmarkMapBool-4      20000000                60.1 ns/op             0 B/op          0 allocs/op
BenchmarkMapBool-4      20000000                59.2 ns/op             0 B/op          0 allocs/op
BenchmarkMapBool-4      20000000                60.1 ns/op             0 B/op          0 allocs/op
BenchmarkMapBool-4      20000000                58.1 ns/op             0 B/op          0 allocs/op
BenchmarkMapBool-4      30000000                59.9 ns/op             0 B/op          0 allocs/op
BenchmarkMapStruct-4    20000000                62.3 ns/op             0 B/op          0 allocs/op
BenchmarkMapStruct-4    20000000                68.4 ns/op             0 B/op          0 allocs/op
BenchmarkMapStruct-4    20000000                62.2 ns/op             0 B/op          0 allocs/op
BenchmarkMapStruct-4    20000000                65.4 ns/op             0 B/op          0 allocs/op
BenchmarkMapStruct-4    20000000                64.5 ns/op             0 B/op          0 allocs/op
```
Again the difference isn't big in both cases, but IMO they nullify and the CPU usage is more important IMO

Reply:
But using struct is not much friendly for user, so I will go with @na––

Reply:
This seems to be resolved then :smile: